Crates_Server_side/services/level.js

```diff
@@ -0,0 +1,163 @@
+    levelMetadataJSON.updatedAt = Date.now();
+    if (!levelMetadataJSON.createdAt) {
+        levelMetadataJSON.createdAt = levelMetadataJSON.updatedAt;
+    }
+    if (!levelMetadataJSON.id) {
+        levelMetadataJSON.id = uuidv1();
+        levelMetadataJSON.createdBy = userId;
+    }
+    return new Promise((resolve, reject) => {
+        s3Service.upload(levelMetadataJSON, (error, data) => {
+            if (error) {
+                let response = {
+                    err: true,
+                    data: error
+                };
+                reject(response);
+            }
+            else {
+                console.log(levelMetadataJSON);
+                console.log("Level Metadata file uploaded successfully at " + data.Location);
+                let response = {
+                    err: false,
+                    data: data
+                };
+                resolve(response);
+            }
+        })
+    });
+};
+
+exports.getLevel = (game, difficulty, createdBy, createdAt) => {
+    return new Promise((resolve, reject) => {
+        s3Service.download(
+            game, 
+            difficulty, 
+            createdBy, 
+            createdAt,
+            (error, data) => {
+                if (error) {
+                    let response = {
+                        err: true,
+                        data: error
+                    };
+                    reject(response);
+                }
+                else {
+                    let levelMetadataJSON = JSON.parse(data.Body.toString());
+                    console.log(levelMetadataJSON);
+                    console.log("Retrieved level Metadata of the level: " + 
+                        s3Service.getLevelKey(game, difficulty, createdBy, createdAt) + " successfully");
+                    let response = {
+                        err: false,
+                        data: levelMetadataJSON
+                    };
+                    resolve(response); 
+                }
+            }
+        );
+    });
+};
+
+exports.getAllLevels = (game) => {
+    return new Promise((resolve, reject) => {
+        try {
+            listAllObjectKeysFromS3Bucket(game + '/levels/').
+            then((allObjectKeysList) => {
+                let allLevelKeysList = listAllLevelKeys(allObjectKeysList);
+                console.log("List of All Level Keys:");
+                console.log(allLevelKeysList);
+                let response = {
+                    err: false,
+                    data: allLevelKeysList
+                };
+                resolve(response); 
+            });
+        } catch (error) {
+            let response = {
+                err: true,
+                data: error
+            };
+            reject(response);
+        }
+    });
+};
+
+exports.getLevelsOfReqDifficulty = (game, difficulty) => {
+    return new Promise((resolve, reject) => {
+        try {
+            listAllObjectKeysFromS3Bucket(game + '/levels/' + difficulty).
```

Comment:
Key name generation should be in one place to make changes easily subsequently

Reply:
Can we have discuss on this offline? Not sure what exactly you are conveying. Do you mean listAllObjectKeysFromS3Bucket should also be in s3Service file?